### PR TITLE
cln: Derive more instances in Hledger.Data.Types to reduce orphans.

### PR DESCRIPTION
**Pull request by Xitian9**
Also clean up some outdated or inaccurate comments.

### CODE REVIEW

File: hledger-lib/Hledger/Data/Amount.hs [hunk 1]
```diff
@@ -247,8 +246,8 @@ amt @@ priceamt = amt{aprice=Just $ TotalPrice priceamt}
 -- Prices are ignored and discarded.
 -- Remember: the caller is responsible for ensuring both amounts have the same commodity.
 similarAmountsOp :: (Quantity -> Quantity -> Quantity) -> Amount -> Amount -> Amount
-similarAmountsOp op !Amount{acommodity=_,  aquantity=q1, astyle=AmountStyle{asprecision=p1}}
-                    !Amount{acommodity=c2, aquantity=q2, astyle=s2@AmountStyle{asprecision=p2}} =
+similarAmountsOp op Amount{acommodity=_,  aquantity=q1, astyle=AmountStyle{asprecision=p1}}
+                    Amount{acommodity=c2, aquantity=q2, astyle=s2@AmountStyle{asprecision=p2}} =
```
Comment (simonmichael) [inline hledger-lib/Hledger/Data/Amount.hs:250]: Just curious, how do you know these are redundant ?

**Comment by simonmichael** [comment]
I think it's causing a warning in master somehow ?
https://github.com/simonmichael/hledger/runs/3391904284?check_suite_focus=true

**Comment by Xitian9** [comment]
I think that warning was caused by 993d0b5c54e364b26fe78a8e6ff62bd626e42c1e, which introduced that import without using it anywhere.

File: hledger-lib/Hledger/Data/Amount.hs (hunk 1 above)
Comment (Xitian9) [reply hledger-lib/Hledger/Data/Amount.hs:250]: hlint says so. Pattern matching is always strict. It needs to put it into whnf to see if the pattern matches.

**Comment by Xitian9** [comment]
I think the situation is that `unlines` is already in Prelude, so importing it from `Data.String.Compat` is redundant.

**Comment by simonmichael** [comment]
Cool, thanks. Sorry about the warning in master, my fault.

**Pull request merged**